helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -64,15 +67,15 @@
 
 public class RoutingTableProvider
     implements ExternalViewChangeListener, InstanceConfigChangeListener, ConfigChangeListener,
-    LiveInstanceChangeListener, CurrentStateChangeListener {
+    LiveInstanceChangeListener, CurrentStateChangeListener, CustomizedViewChangeListener {
   private static final Logger logger = LoggerFactory.getLogger(RoutingTableProvider.class);
   private static final long DEFAULT_PERIODIC_REFRESH_INTERVAL = 300000L; // 5 minutes
-  private final AtomicReference<RoutingTable> _routingTableRef;
+  private final Map<String, AtomicReference<RoutingTable>> _routingTableRef;
   private final HelixManager _helixManager;
-  private final RouterUpdater _routerUpdater;
-  private final PropertyType _sourceDataType;
+  private final Map<String, RouterUpdater> _routerUpdater;
```

Comment:
Better to be one updater fand one cache inside.

Reply:
Done.